### PR TITLE
sozo: add dry-run mode

### PR DESCRIPTION
**Pull request by remybar**
Related to https://github.com/dojoengine/dojo/issues/1468.
closes DOJ-129

In dry-run mode, display the computed migration strategy instead of executing it.

### CODE REVIEW

File: crates/sozo/ops/src/migration/mod.rs [hunk 1]
```diff
@@ -683,3 +688,33 @@ pub fn handle_artifact_error(ui: &Ui, artifact_path: &Path, error: anyhow::Error
          --artifacts` to clean artifacts only.\nThen, rebuild your project with `sozo build`."
     )
 }
+
+pub fn print_strategy(ui: &Ui, strategy: &MigrationStrategy) {
+    ui.print("\n📋 Migration Strategy\n");
+
+    if let Some(base) = &strategy.base {
+        ui.print_header("# Base Contract");
+        ui.print_sub(format!("declare (classhash: {:#x})\n", base.diff.local));
```
Comment (glihm) [inline crates/sozo/ops/src/migration/mod.rs:697]:
```suggestion
        ui.print_sub(format!("declare (class hash: {:#x})\n", base.diff.local));
```

File: crates/sozo/ops/src/migration/mod.rs [hunk 2]
```diff
@@ -683,3 +688,33 @@ pub fn handle_artifact_error(ui: &Ui, artifact_path: &Path, error: anyhow::Error
...
+    }
+
+    if let Some(world) = &strategy.world {
+        ui.print_header("# World");
+        ui.print_sub(format!("declare (classhash: {:#x})\n", world.diff.local));
```
Comment (glihm) [inline crates/sozo/ops/src/migration/mod.rs:702]:
```suggestion
        ui.print_sub(format!("declare (class hash: {:#x})\n", world.diff.local));
```

File: crates/sozo/ops/src/migration/mod.rs [hunk 3]
```diff
@@ -683,3 +688,33 @@ pub fn handle_artifact_error(ui: &Ui, artifact_path: &Path, error: anyhow::Error
...
+    }
+
+    if !&strategy.models.is_empty() {
+        ui.print_header(format!("# Models ({})", &strategy.models.len()));
+        for m in &strategy.models {
+            ui.print_sub(format!("register {} (classhash: {:#x})", m.diff.name, m.diff.local));
```
Comment (glihm) [inline crates/sozo/ops/src/migration/mod.rs:708]:
```suggestion
            ui.print_sub(format!("register {} (class hash: {:#x})", m.diff.name, m.diff.local));
```

File: crates/sozo/ops/src/migration/mod.rs [hunk 4]
```diff
@@ -683,3 +688,33 @@ pub fn handle_artifact_error(ui: &Ui, artifact_path: &Path, error: anyhow::Error
...
+        }
+        ui.print(" ");
+    }
+
+    if !&strategy.contracts.is_empty() {
+        ui.print_header(format!("# Contracts ({})", &strategy.contracts.len()));
+        for c in &strategy.contracts {
+            ui.print_sub(format!("deploy {} (classhash: {:#x})", c.diff.name, c.diff.local));
```
Comment (glihm) [inline crates/sozo/ops/src/migration/mod.rs:716]:
```suggestion
            ui.print_sub(format!("deploy {} (class hash: {:#x})", c.diff.name, c.diff.local));
```

**Review by glihm — COMMENTED** [review]
Good first pass @remybar!

I'm wondering in which measure we're able to notify the user if the contract (last part of the strategy) will be upgraded in place, or actually deployed.

For the world, it's not yet upgradable during migration, so it's fine for now having it like so. :+1:

**Comment by remybar** [comment]
> Good first pass @remybar!
> 
> I'm wondering in which measure we're able to notify the user if the contract (last part of the strategy) will be upgraded in place, or actually deployed.

Yes, good idea 👍

File: crates/sozo/ops/src/migration/mod.rs [hunk 5]
```diff
@@ -683,3 +692,78 @@ pub fn handle_artifact_error(ui: &Ui, artifact_path: &Path, error: anyhow::Error
...
+            .call(
+                FunctionCall {
+                    contract_address: world_address,
+                    calldata: vec![],
+                    entry_point_selector: get_selector_from_name("base").unwrap(),
+                },
+                BlockId::Tag(BlockTag::Pending),
+            )
+            .await
+        {
```
Comment (glihm) [inline crates/sozo/ops/src/migration/mod.rs:715]: @lambda-0x just to keep a note aside, after the modification on upgrade for the world + outputting the base class hash in the manifest, we shouldn't fetch, but only take the base class hash from the local manifest. :+1:

**Review by glihm — APPROVED** [review]
@remybar thank you again for the work done here! :rocket:

**Comment by tarrencev** [comment]
Not something for this PR but what do you guys think of having:

sozo migrate plan (generate migration plan)
sozo migrate apply (apply migration plan)

**Comment by remybar** [comment]
> Not something for this PR but what do you guys think of having:
> 
> sozo migrate plan (generate migration plan) sozo migrate apply (apply migration plan)

Yes, it might be clearer to use some sub-commands rather than options for this 👍

**Comment by lambda-0x** [comment]
> Not something for this PR but what do you guys think of having:
> 
> sozo migrate plan (generate migration plan) sozo migrate apply (apply migration plan)

on high level that makes sense, `migrate plan` would be similar to passing `--dry-run` but instead of just printing information it would write to a file all the changes that it will apply and `migrate apply` would read that file and actually do the operation

**Pull request merged**